libraries/GDBStub/src/internal/gdbstub.c

```diff
@@ -233,73 +175,157 @@ static long ATTR_GDBFN gdbGetHexVal(unsigned char **ptr, int bits) {
 }
 
 //Swap an int into the form gdb wants it
-static int ATTR_GDBFN iswap(int i) {
-	int r;
-	r=((i>>24)&0xff);
-	r|=((i>>16)&0xff)<<8;
-	r|=((i>>8)&0xff)<<16;
-	r|=((i>>0)&0xff)<<24;
-	return r;
+static int /*ATTR_GDBFN*/ iswap(int i) {
+	return ((i >> 24) & 0xff)
+		| (((i >> 16) & 0xff) << 8)
+		| (((i >> 8) & 0xff) << 16)
+		| (((i >> 0) & 0xff) << 24);
 }
 
 //Read a byte from the ESP8266 memory.
-static unsigned char ATTR_GDBFN readbyte(unsigned int p) {
-	int *i=(int*)(p&(~3));
-	if (p<0x20000000 || p>=0x60000000) return -1;
-	return *i>>((p&3)*8);
+static unsigned char /*ATTR_GDBFN*/ readbyte(unsigned int p) {
+	if (p < 0x20000000 || p >= 0x60000000) return -1;
+	int *i = (int*)(p & ~3);
+	return *i >> ((p & 3) * 8);
 }
 
 //Write a byte to the ESP8266 memory.
-static void ATTR_GDBFN writeByte(unsigned int p, unsigned char d) {
-	int *i=(int*)(p&(~3));
-	if (p<0x20000000 || p>=0x60000000) return;
-	if ((p&3)==0) *i=(*i&0xffffff00)|(d<<0);
-	if ((p&3)==1) *i=(*i&0xffff00ff)|(d<<8);
-	if ((p&3)==2) *i=(*i&0xff00ffff)|(d<<16);
-	if ((p&3)==3) *i=(*i&0x00ffffff)|(d<<24);
+static void /*ATTR_GDBFN*/ writeByte(unsigned int p, unsigned char d) {
+	if (p < 0x20000000 || p >= 0x60000000) return;
+	int *i = (int*)(p & ~3);
+	if ((p & 3) == 0) *i = (*i & 0xffffff00) | (d << 0);
+	else if ((p & 3) == 1) *i = (*i & 0xffff00ff) | (d << 8);
+	else if ((p & 3) == 2) *i = (*i & 0xff00ffff) | (d << 16);
+	else if ((p & 3) == 3) *i = (*i & 0x00ffffff) | (d << 24);
 }
 
 //Returns 1 if it makes sense to write to addr p
-static int ATTR_GDBFN validWrAddr(int p) {
-	if (p>=0x3ff00000 && p<0x40000000) return 1;
-	if (p>=0x40100000 && p<0x40140000) return 1;
-	if (p>=0x60000000 && p<0x60002000) return 1;
-	return 0;
+static int /*ATTR_GDBFN*/ validWrAddr(int p) {
+	return (p >= 0x3ff00000 && p < 0x40000000)
+		|| (p >= 0x40100000 && p < 0x40140000)
+		|| (p >= 0x60000000 && p < 0x60002000);
 }
 
-/*
-Register file in the format lx106 gdb port expects it.
-Inspired by gdb/regformats/reg-xtensa.dat from
-https://github.com/jcmvbkbc/crosstool-NG/blob/lx106-g%2B%2B/overlays/xtensa_lx106.tar
-As decoded by Cesanta.
-*/
-struct regfile {
-	uint32_t a[16];
-	uint32_t pc;
-	uint32_t sar;
-	uint32_t litbase;
-	uint32_t sr176;
-	uint32_t sr208;
-	uint32_t ps;
-};
+
+static inline bool ATTR_GDBFN gdbRxFifoIsEmpty() {
```

Comment:
This whole FIFO code looks very similar to what's in the uart.* files. Is it worth investigating unification?

Reply:
Not worth it in my opinion after going through it a few times.

The GDB infrastructure is completely independent from the `struct uart_` stuff in `uart.c` in order to simplify the design (i.e. user can free a `uart_` structure by doing a Serial.end(). but GDB still needs to continue to run).

This also minimizes the logic changes in `uart.c` (an already hairy piece of code!) so it's also got that as a positive.